### PR TITLE
Port mse_lose to ATen

### PR DESCRIPTION
**Pull request by XiaobingSuper**
@VitalyFedyunin, This PR is about port mse lose to Aten：

**Test script:**
```
import torch
import torch.nn as nn
import time

def _time():
    if torch.cuda.is_available():
        torch.cuda.synchronize()
    return time.time()

device = "cpu"
loss = nn.MSELoss(reduction = 'sum')
if torch.cuda.is_available():
    device = "cuda"
    loss = loss.cuda()

#warm up
for n in [100, 10000]:
    input = torch.randn(128, n, requires_grad=True, device=device)
    target = torch.randn(128, n, device=device)
    for i in range(1000):
        output = loss(input, target)
        output.backward()

#get running time
for n in [100, 10000]:
    fwd_t = 0
    bwd_t = 0
    input = torch.randn(128, n, requires_grad=True, device=device)
    target = torch.randn(128, n, device=device)
    for i in range(10000):
        t1 = _time()
        output = loss(input, target)
        t2 = _time()
        output.backward()
        t3 = _time()
        fwd_t = fwd_t + (t2 -t1)
        bwd_t = bwd_t + (t3 - t2)
    fwd_avg = fwd_t / 10000 * 1000
    bwd_avg = bwd_t / 10000 * 1000
    print("input size(128, %d) forward time is %.2f (ms); backwad avg time is %.2f (ms)."
          % (n, fwd_avg, bwd_avg))
```
**Test Device:** CPU: skx-8180, GPU: Tesla P40.

### Perfromance:

**Before:**
```
GPU:
reduction=’mean’
input size(128, 100) forward time is 0.08 (ms); backwad avg time is 0.14 (ms).
input size(128, 10000) forward time is 0.12 (ms); backwad avg time is 0.21 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.09 (ms); backwad avg time is 0.15 (ms).
input size(128, 10000) forward time is 0.11 (ms); backwad avg time is 0.20 (ms).

CPU:
OMP_NUM_THREADS=56
reduction=’mean’
input size(128, 100) forward time is 0.06 (ms); backwad avg time is 0.09 (ms).
input size(128, 10000) forward time is 3.49 (ms); backwad avg time is 3.23 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.06 (ms); backwad avg time is 0.09 (ms).
input size(128, 10000) forward time is 3.49 (ms); backwad avg time is 3.23 (ms).

OMP_NUM_THREADS=1
reduction=’mean’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.04 (ms).
input size(128, 10000) forward time is 1.41 (ms); backwad avg time is 1.66 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.04 (ms).
input size(128, 10000) forward time is 1.44 (ms); backwad avg time is 1.68 (ms).
```

**After:**
```
GPU:
reduction=’mean’
input size(128, 100) forward time is 0.07 (ms); backwad avg time is 0.13 (ms).
input size(128, 10000) forward time is 0.13 (ms); backwad avg time is 0.20 (ms).

reduction=’sum’
input size(128, 100) forward time is 0.07 (ms); backwad avg time is 0.14 (ms).
input size(128, 10000) forward time is 0.13 (ms); backwad avg time is 0.20 (ms).

CPU:
OMP_NUM_THREADS=56
reduction=’mean’
input size(128, 100) forward time is 0.04 (ms); backwad avg time is 0.06 (ms).
input size(128, 10000) forward time is 0.14 (ms); backwad avg time is 0.30 (ms).

reduction=’sum’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.06 (ms).
input size(128, 10000) forward time :qis 0.13 (ms); backwad avg time is 0.30 (ms).

OMP_NUM_THREADS=1
reduction=’mean’
input size(128, 100) forward time is 0.04 (ms); backwad avg time is 0.05 (ms).
input size(128, 10000) forward time is 0.85 (ms); backwad avg time is 1.27 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.04 (ms).
input size(128, 10000) forward time is 0.83 (ms); backwad avg time is 1.27 (ms).
```

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Can you please add benchmarks with `OMP_NUM_THREADS=1`?

Also GPU numbers looks concerning to me, I will check code accurately for the slow down reason.

**Comment by VitalyFedyunin** [comment]
#24598 Migrate mse_loss from the TH to Aten (CUDA)
#24599 Migrate mse_loss_backward from the TH to Aten (CUDA)
#24732 Migrate mse_loss from the TH to Aten (CPU)
#24733 Migrate mse_loss_backward from the TH to Aten (CPU)

**Comment by XiaobingSuper** [comment]
The following is the perfomance data for `OMP_NUM_THREADS=1`,  the run script is :
```
num_threads=$1
script=$2
last_core=`expr $num_threads - 1`


echo "using $num_threads OMP threads"
echo "bind cores to 0~$last_core"

export OMP_NUM_THREADS=$num_threads
export KMP_AFFINITY=granularity=fine,compact,1,0

numactl --physcpubind=0-$last_core --membind=0 python $script
```

How to run?  `./run.sh 1 mse_cpu.py`

Before:
```
reduction=’mean’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.04 (ms).
input size(128, 10000) forward time is 1.52 (ms); backwad avg time is 1.83 (ms).

reduction=’sum’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.04 (ms).
input size(128, 10000) forward time is 1.53 (ms); backwad avg time is 1.85 (ms).
```
After:
```
reduction=’mean’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.05 (ms).
input size(128, 10000) forward time is 1.04 (ms); backwad avg time is 1.68 (ms).

reduction=’sum’
input size(128, 100) forward time is 0.02 (ms); backwad avg time is 0.04 (ms).
input size(128, 10000) forward time is 1.04 (ms); backwad avg time is 1.69 (ms).
```

**Comment by XiaobingSuper** [comment]
I will also see the reason of GPU performance degradation. Thanks!

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, do you any finding about the GPU degration? Using same inputs, I test the toch.sub computer time, this time is about 70% of origin mse_loss compute time, perharps there has big overhead when port sub to Aten(https://github.com/pytorch/pytorch/pull/8919). So do we compare the aten sub with THC sub for large input size?

**Comment by VitalyFedyunin** [comment]
You need to fuse operations to get comparable performance (CPU and CUDA).

File: aten/src/ATen/native/Loss.cpp [hunk 1]
```diff
@@ -148,4 +148,36 @@ Tensor poisson_nll_loss(const Tensor& input, const Tensor& target, const bool lo
...
+Tensor& mse_loss_out(Tensor&result, const Tensor& input, const Tensor& target, int64_t reduction) {
+  if (reduction != Reduction::None) {
+    auto loss = input.sub(target).pow_(2);
+    if (reduction == Reduction::Mean) {
+      at::mean_out(result, loss, 0);
+    } else {
+      at::sum_out(result, loss, 0);
+    }
+  } else {
+    at::sub_out(result, input, target).pow_(2);
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Loss.cpp]: As example this is 2 separate CPU loops or CUDA kernels, one is for `sub`, another is for `pow`.

File: aten/src/THCUNN/generic/MSECriterion.cu [hunk 2]
```diff
@@ -1,126 +0,0 @@
...
-
-    THCTensor_(free)(state, input);
-    THCTensor_(free)(state, target);
-
-    THCTensor_(set0d)(state, output, ScalarConvert<accreal, scalar_t>::to(sum));
-    return;
-  }
-
-  THCTensor_(resizeAs)(state, output, input);
-  THC_pointwiseApply3<scalar_t, scalar_t, scalar_t>(
```
Comment (VitalyFedyunin) [inline aten/src/THCUNN/generic/MSECriterion.cu]: This CUDA loop does `sub` and `pow` inside same kernel.

File: aten/src/THNN/generic/MSECriterion.c [hunk 3]
```diff
@@ -1,68 +0,0 @@
...
-
-    if (reduction == Reduction::Mean)
-      sum /= THTensor_(nElement)(input);
-
-    THTensor_(set0d)(output, (scalar_t)sum);
-    return;
-  }
-
-  THTensor_(resizeAs)(output, input);
-  TH_TENSOR_APPLY3(scalar_t, input, scalar_t, target, scalar_t, output,
```
Comment (VitalyFedyunin) [inline aten/src/THNN/generic/MSECriterion.c]: This CPU implementation does `sub` and `pow` inside of single loop.

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
You need to fuse operations together to get similar/better performance.

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin @zou3519, the backward of [L1_loss](https://github.com/pytorch/pytorch/pull/26795) and mse_loss also have performance overhead which call sub and mul in two loops, there two method to fuse them: using TensorIterator to implement a ternary kenel: sub_mul, or just use parallel_for, which one is acceptabled to you? Thanks!

File: aten/src/ATen/native/Loss.cpp [hunk 4]
```diff
@@ -148,4 +148,36 @@ Tensor poisson_nll_loss(const Tensor& input, const Tensor& target, const bool lo
 
     return apply_loss_reduction(loss, reduction);
 }
+
+Tensor mse_loss(const Tensor& input, const Tensor& target, int64_t reduction) {
```
Comment (zou3519) [inline aten/src/ATen/native/Loss.cpp]: https://github.com/pytorch/pytorch/blob/8fb756d3b25be6b8f56bbd0d7b52b1f1ed1e7119/torch/nn/functional.py#L2183-L2205

I think we can delete the `broadcast_tensors` call now after this change.

**Comment by XiaobingSuper** [comment]
The following is the performance after fusing operations:
```
GPU:
reduction=’mean’
input size(128, 100) forward time is 0.07 (ms); backwad avg time is 0.14 (ms).
input size(128, 10000) forward time is 0.13 (ms); backwad avg time is 0.20 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.07 (ms); backwad avg time is 0.13 (ms).
input size(128, 10000) forward time is 0.13 (ms); backwad avg time is 0.20 (ms).

CPU:
reduction=’mean’
input size(128, 100) forward time is 0.04 (ms); backwad avg time is 0.05 (ms).
input size(128, 10000) forward time is 0.13 (ms); backwad avg time is 0.15 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.04 (ms); backwad avg time is 0.06 (ms).
input size(128, 10000) forward time is 0.12 (ms); backwad avg time is 0.14 (ms).

OMP_NUM_THREADS=1
reduction=’mean’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.04 (ms).
input size(128, 10000) forward time is 0.85 (ms); backwad avg time is 1.26 (ms).
reduction=’sum’
input size(128, 100) forward time is 0.03 (ms); backwad avg time is 0.04 (ms).
input size(128, 10000) forward time is 0.83 (ms); backwad avg time is 1.27 (ms).
```
After using fused operators, the overhead decreased to 0.02 ms for GPU forward, but there has performance improvement for other cases.

File: aten/src/ATen/native/Loss.cpp [hunk 5]
```diff
@@ -148,4 +154,48 @@ Tensor poisson_nll_loss(const Tensor& input, const Tensor& target, const bool lo
...
+}
+
+Tensor mse_loss_backward(const Tensor& grad_output, const Tensor& input, const Tensor& target, int64_t reduction) {
+  Tensor grad_input = at::zeros_like(input);
+  return at::mse_loss_backward_out(grad_input, grad_output, input, target, reduction);
+}
+
+Tensor& mse_loss_backward_out(Tensor& grad_input, const Tensor& grad_output,
+    const Tensor& input, const Tensor& target, int64_t reduction) {
+  auto norm = reduction == Reduction::Mean ? 2. / input.numel() : 2.;
```
Comment (XiaobingSuper) [inline aten/src/ATen/native/Loss.cpp]: @zou3519 , there have a problem for mse_backward if delete **broadcast_tensors**,  for input.size=(2, 1), target.size =(2,10), the norm number should be 20, but get 2 in there.  So I didn't remove this **broadcast_tensors** method now.

Comment (zou3519) [reply aten/src/ATen/native/Loss.cpp]: Oh, I see... It's because we have a custom backward implementation for it. That's fine for now then, we can work on improving `mse_loss` separately after the port, thank you for pointing that out.

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin @zou3519

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, please help review.

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, Can you find some people to review the code, I have rebased it too many times. :disappointed:

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, please help review the code, thanks!

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by XiaobingSuper** [comment]
> Looks good, can you please update PR descriptions with most recent benchmark numbers.

Benchmark numbers are updated, thanks!

**Review by VitalyFedyunin — APPROVED** [review]
